File: qqbot-domain/src/test/java/com/mieai/qqbot/domain/bot/GatewayIntentsTest.java
package com.mieai.qqbot.domain.bot;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
import static org.assertj.core.api.Assertions.assertThatNullPointerException;

import org.junit.jupiter.api.Test;

class GatewayIntentsTest {
    @Test
    void exposesTheCurrentOfficialIntentBitsAndSafeDefault() {
        assertThat(GatewayIntents.GUILDS.bits()).isEqualTo(1L << 0);
        assertThat(GatewayIntents.GUILD_MEMBERS.bits()).isEqualTo(1L << 1);
        assertThat(GatewayIntents.GUILD_MESSAGES.bits()).isEqualTo(1L << 9);
        assertThat(GatewayIntents.GUILD_MESSAGE_REACTIONS.bits()).isEqualTo(1L << 10);
        assertThat(GatewayIntents.DIRECT_MESSAGE.bits()).isEqualTo(1L << 12);
        assertThat(GatewayIntents.GROUP_AND_C2C_EVENT.bits()).isEqualTo(1L << 25);
        assertThat(GatewayIntents.INTERACTION.bits()).isEqualTo(1L << 26);
        assertThat(GatewayIntents.MESSAGE_AUDIT.bits()).isEqualTo(1L << 27);
        assertThat(GatewayIntents.FORUMS_EVENT.bits()).isEqualTo(1L << 28);
        assertThat(GatewayIntents.AUDIO_ACTION.bits()).isEqualTo(1L << 29);
        assertThat(GatewayIntents.PUBLIC_GUILD_MESSAGES.bits()).isEqualTo(1L << 30);
        assertThat(GatewayIntents.DEFAULT_MESSAGE_EVENTS)
                .isSameAs(GatewayIntents.GROUP_AND_C2C_EVENT);
    }

    @Test
    void representsNoIntentsWithCanonicalInstance() {
        assertThat(GatewayIntents.of(0L)).isSameAs(GatewayIntents.NONE);
        assertThat(GatewayIntents.NONE.isEmpty()).isTrue();
    }

    @Test
    void combinesRemovesAndChecksIntentMasks() {
        GatewayIntents messages = GatewayIntents.of(1L << 9);
        GatewayIntents interactions = GatewayIntents.of(1L << 26);

        GatewayIntents combined = messages.union(interactions);

        assertThat(combined.bits()).isEqualTo((1L << 9) | (1L << 26));
        assertThat(combined.containsAll(messages)).isTrue();
        assertThat(combined.containsAll(interactions)).isTrue();
        assertThat(messages.containsAll(interactions)).isFalse();
        assertThat(combined.without(messages)).isEqualTo(interactions);
    }

    @Test
    void preservesUnknownNonNegativeBits() {
        GatewayIntents intents = GatewayIntents.of(1L << 60);

        assertThat(intents.bits()).isEqualTo(1L << 60);
    }

    @Test
    void rejectsNegativeMaskAndNullOperands() {
        assertThatIllegalArgumentException().isThrownBy(() -> new GatewayIntents(-1L));
        assertThatNullPointerException().isThrownBy(() -> GatewayIntents.NONE.containsAll(null));
        assertThatNullPointerException().isThrownBy(() -> GatewayIntents.NONE.union(null));
        assertThatNullPointerException().isThrownBy(() -> GatewayIntents.NONE.without(null));
    }
}

File: qqbot-domain/src/test/kotlin/com/mieai/qqbot/domain/bot/GatewayIntentsTest.kt
package com.mieai.qqbot.domain.bot

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatIllegalArgumentException
import org.junit.jupiter.api.Test

class GatewayIntentsTest {
    @Test
    fun exposesTheCurrentOfficialIntentBitsAndSafeDefault() {
        assertThat(GatewayIntents.GUILDS.bits).isEqualTo(1L shl 0)
        assertThat(GatewayIntents.GUILD_MEMBERS.bits).isEqualTo(1L shl 1)
        assertThat(GatewayIntents.GUILD_MESSAGES.bits).isEqualTo(1L shl 9)
        assertThat(GatewayIntents.GUILD_MESSAGE_REACTIONS.bits).isEqualTo(1L shl 10)
        assertThat(GatewayIntents.DIRECT_MESSAGE.bits).isEqualTo(1L shl 12)
        assertThat(GatewayIntents.GROUP_MEMBERS.bits).isEqualTo(1L shl 24)
        assertThat(GatewayIntents.GROUP_AND_C2C_EVENT.bits).isEqualTo(1L shl 25)
        assertThat(GatewayIntents.INTERACTION.bits).isEqualTo(1L shl 26)
        assertThat(GatewayIntents.MESSAGE_AUDIT.bits).isEqualTo(1L shl 27)
        assertThat(GatewayIntents.FORUMS_EVENT.bits).isEqualTo(1L shl 28)
        assertThat(GatewayIntents.AUDIO_ACTION.bits).isEqualTo(1L shl 29)
        assertThat(GatewayIntents.PUBLIC_GUILD_MESSAGES.bits).isEqualTo(1L shl 30)
        assertThat(GatewayIntents.DEFAULT_MESSAGE_EVENTS).isSameAs(GatewayIntents.GROUP_AND_C2C_EVENT)
    }

    @Test
    fun representsNoIntentsWithCanonicalInstance() {
        assertThat(GatewayIntents.of(0L)).isSameAs(GatewayIntents.NONE)
        assertThat(GatewayIntents.NONE.isEmpty()).isTrue()
    }

    @Test
    fun combinesRemovesAndChecksIntentMasks() {
        val messages = GatewayIntents.of(1L shl 9)
        val interactions = GatewayIntents.of(1L shl 26)
        val combined = messages.union(interactions)

        assertThat(combined.bits).isEqualTo((1L shl 9) or (1L shl 26))
        assertThat(combined.containsAll(messages)).isTrue()
        assertThat(combined.containsAll(interactions)).isTrue()
        assertThat(messages.containsAll(interactions)).isFalse()
        assertThat(combined.without(messages)).isEqualTo(interactions)
    }

    @Test
    fun preservesUnknownNonNegativeBits() {
        val intents = GatewayIntents.of(1L shl 60)
        assertThat(intents.bits).isEqualTo(1L shl 60)
    }

    @Test
    fun rejectsNegativeMask() {
        assertThatIllegalArgumentException().isThrownBy { GatewayIntents(-1L) }
    }
}
